icu4c/source/common/uloc.cpp

```diff
@@ -1184,104 +1184,132 @@ _getLanguage(const char* localeID,
         localeID += 3;
     }
 
+    constexpr int32_t MAXLEN = ULOC_LANG_CAPACITY - 1;  // Minus NUL.
+
     /* if it starts with i- or x- then copy that prefix */
-    if(_isIDPrefix(localeID)) {
-        result.append((char)uprv_tolower(*localeID), status);
-        result.append('-', status);
-        localeID+=2;
+    int32_t len = _isIDPrefix(localeID) ? 2 : 0;
+    while (!_isTerminator(localeID[len]) && !_isIDSeparator(localeID[len])) {
+        len++;
     }
+    if (len > MAXLEN) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return;
+    }
+
+    *pEnd = localeID + len;
+    if (sink == nullptr || len == 0) { return; }
 
-    /* copy the language as far as possible and count its length */
-    while(!_isTerminator(*localeID) && !_isIDSeparator(*localeID)) {
-        result.append((char)uprv_tolower(*localeID), status);
-        localeID++;
+    char scratch[MAXLEN];
+    int32_t capacity = 0;
+    int32_t min_capacity = uprv_min(len, 4);  // Minimum 3 letters plus NUL.
+    char* buffer = sink->GetAppendBuffer(
+            min_capacity, MAXLEN, scratch, UPRV_LENGTHOF(scratch), &capacity);
+    U_ASSERT(capacity >= min_capacity);
```

Comment:
```suggestion
    int32_t minCapacity = uprv_min(len, 4);  // Minimum 3 letters plus NUL.
    char* buffer = sink->GetAppendBuffer(
            minCapacity, minCapacity, scratch, UPRV_LENGTHOF(scratch), &capacity);
    U_ASSERT(capacity >= minCapacity);
```

Reply:
Why would this be better?

Reply:
1. lowerCamelCase as usual in ICU
2. don't need to ask for more space than we plan to write when the sink does have to reallocate

Reply:
1. The variable named `min_capacity` is being passed to the parameter named `min_capacity` in `GetAppendBuffer()`, there's a nice consistency in that, that can't be important to change.
2. I don't think I understand the changes you propose here at all. Let's take a concrete example:

The first entry in `LANGUAGES_3` is `"aar"` which should be shortened to `"aa"`. With the change you requested before, the value of `min_capacity` will be `uprv_min(len, 4)` which will be 3 which is one less than needed in order to store `"aar"` plus a NUL terminator in the buffer. I had previously assumed that this would work because the value of `desired_capacity_hint` was larger and the scratch buffer was large enough to accomodate that larger value. But when I now read the implementation of `CheckedArrayByteSink::GetAppendBuffer()` I see that it ignores the value of `desired_capacity_hint` and will return a buffer of size 3 if that's all that's available in the underlying array, which means that the NUL terminator written later will then write out of bounds. That's not good.

Did you maybe mean that the buffer here should be allocated with `min_capacity` set to `uprv_max(len, 4)` so that it allocates at least enough space to do the three-letter lookup but not any more than that unless needed?

That would make sense to me, but in that case the corresponding changes in `_getRegion()` don't make any sense at all, the hardcoded size in the first iteration of this PR is the smallest buffer that could be used.

So should I change the code here to use `uprv_max(len, 4)` and revert the changes in `_getRegion()`, or am I missing something?

Reply:
1. The StringPiece and ByteSink classes came from Google with its different conventions. In code _using_ them we normally use ICU conventions.
2. Snafu -- need max, not min. Sorry!
3. If we don't need a max because the buffer has a fixed length, then that's fine. Note that I didn't comment on using LENGTH for scripts; and I didn't look at the numeric values of the other constants.

Reply:
OK, I've now fixed up the code in these functions to always allocate the minimal buffer size needed, but not more than that for language tags shorter than the longest possible language tag.

Reply:
thank you for doing what i meant not what i said!! :-)